### PR TITLE
Static lighting and stuff

### PR DESCRIPTION
**Pull request by guycalledfrank**
Many things are here.

I. Static lights. Set model.isStatic=true and light.isStatic=true to enable. These lights will only affect the triangles they intersect. The benefits are:
- You can have large amounts of lights in the scene without being afraid to go out of uniforms;
- It's obviously faster than lighting every pixel with every light;
- Doesn't use as much memory as lightmaps;
- Fully correct specular;
- You can dynamically change lights' colors or shadow/cookie properties. You can also dynamically narrow the radius or the spotlight cone;
- Objects can cast dynamic shadows inside static lights;
- In fact, you CAN move static lights and the associated geometry (that makes me unsure about the "isStatic" name). As long, as lights intersect same triangles, it will work (so for example, you can have a moving self-lit train).

You can recalculate static lighting data by calling renderer.prepareStaticMeshes(device, scene). It's usually rather fast (60ms for the space demo). This function is called automatically on app launch.
Also a bunch of new profiler stats are added.
Limitation: can't have more than 32 static lights touching one draw call, because of some optimizations. It's possible to remove these optimizations, but... I'd rather have them.

II. Directional lightmaps. Set light.bakeDir=true, then call lightmapper.bake(null, pc.BAKE_COLORDIR) to get it (it's also the new default value). Renders average light direction along with a color lightmap. The light direction map allows to reconstruct bump-mapped lighting and specular. The cost is 2 texture reads + one light calculation + tiny other stuff. Intersecting many lights with bakeDir=true may lead to incorrect bump/specular. As long, as lights don't intersect much, you can bake any amount of them.

III. Cookie transformations. Only work for spotlights (to rotate cubemap cookie, it's enough to just rotate the node, shadows ignore pointlight rotation). The new properties are:
- cookieAngle (float): rotation angle;
- cookieScale (vec2): scaling;
- cookieOffset (vec2): translation offset.

Additional:
- light.setCastShadows is optimized;
- fixed incorrect lightmap dilation https://github.com/playcanvas/engine/issues/695 ;
- fixed some bugs in StandardMaterial's key generation;
- fixed some frustum culling bug which was related to parent scale;
- added light.getBoundingBox();
- added docs for pc.BoundingBox#add;
- added docs for lightmapper.

### CODE REVIEW

File: src/framework/components/light/component.js [hunk 1]
```diff
@@ -87,6 +91,10 @@ pc.extend(pc, function () {
      * @property {Number} cookieIntensity Projection texture intensity (default is 1).
      * @property {Boolean} cookieFalloff Toggle normal spotlight falloff when projection texture is used. When set to false, spotlight will work like a pure texture projector (only fading with distance). Default is false.
      * @property {String} cookieChannel  Color channels of the projection texture to use. Can be "r", "g", "b", "a", "rgb" or any swizzled combination.
+     * @property {Number} cookieAngle Angle for spotlight cookie rotation.
+     * @property {pc.Vec2} cookieScale Spotlight cookie scale.
+     * @property {pc.Vec2} cookieOffset Spotlight cookie position offset.
+     * @property {Boolean} isStatic Mark light as non-movable (optimization)
```
Comment (willeastcott) [inline src/framework/components/light/component.js]: ...'with respect to the mesh instances it lights'?

**Comment by vkalpias** [comment]
Should we rename isStatic to static?

File: src/framework/components/light/component.js (hunk 1 above)
Comment (guycalledfrank) [inline src/framework/components/light/component.js]: It means it will only affect the triangles of static models it intersects with

**Comment by guycalledfrank** [comment]
>Should we rename isStatic to static? 
I wanted to, but then "static" is a reserved keyword in JS. Although It SHOULD work as a variable name, but... I just don't like the idea, I think some day it can break - maybe in browser, maybe in our build tool, maybe somewhere else. Also Sublime annoyingly highlights "static".

File: src/framework/components/light/component.js (hunk 1 above)
Comment (willeastcott) [inline src/framework/components/light/component.js]: I know. I'm trying to suggest an edit to also explain that static doesn't completely mean static (as with the moving train example).

Comment (guycalledfrank) [inline src/framework/components/light/component.js]: Hmmm... then we also need a good correction to this: https://github.com/playcanvas/engine/pull/697/files#diff-a01b339213b1ec54098b0d4161e1bceeR28

**Pull request merged**